core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,550 @@
+/*
+ * Copyright (C) 2020-2021 Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in
+ *      the documentation and/or other materials provided with the
+ *      distribution.
+ *   3. Neither the name of Intel Corporation nor the names of its
+ *      contributors may be used to endorse or promote products derived
+ *      from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "enclave_hsm_t.h"
+#include "log_utils.h"
+#include "sgx_tseal.h"
+
+#include <string>
+#include <stdio.h>
+#include <stdbool.h>
+#include <mbusafecrt.h>
+
+#include "sgx_report.h"
+#include "sgx_utils.h"
+#include "sgx_tkey_exchange.h"
+
+#include "datatypes.h"
+#include "openssl/rsa.h"
+#include "openssl/evp.h"
+#include "openssl/ec.h"
+#include "openssl/pem.h"
+#include "openssl/bio.h"
+#include "openssl/rand.h"
+#include "openssl/err.h"
+
+#include "key_factory.h"
+#include "key_operation.h"
+
+#define SGX_DOMAIN_KEY_SIZE 16
+
+#define RSA_2048_KEY_BITS 2048
+#define RSA_3072_KEY_BITS 3072
+#define RSA_4096_KEY_BITS 4096
+
+#define ECC_MAX_PLAINTEXT_SIZE 256
+
+sgx_aes_gcm_128bit_key_t g_domain_key = {0};
+
+using namespace std;
+
+uint32_t ehsm_calc_keyblob_len(const uint32_t payload_size)
+{
```

Comment:
this function is designed to calculate the keyblob length for each keyspec, I don't see you followed it..

even the RSA/ECC keyblob length should be calculated here though they are hardcoded(since you can't get the exactly length for them.)

Reply:
motify the function by keyspec